src/passes/OptimizeInstructions.cpp

```diff
@@ -917,6 +917,42 @@ struct OptimizeInstructions
     if (binary->right->is<Const>()) {
       return;
     }
+    // Prefer subexpressions with constants on the right as well.
+    if (auto* left = binary->left->dynCast<Binary>()) {
+      if (left->right->is<Const>()) {
+        bool shouldSwap = true;
+        // don't swap if another expression also contain constant
+        // which equal to -1 or 0
```

Comment:
But again, why not do it with 2 or 3 or other constants too? It wouldn't hurt. And the more canonicalization, the simpler things are overall.

Special-casing 0 and -1 would require explaining exactly why they are special here. Otherwise, the code is confusing to read I think.

Reply:
Concretely, how about doing this?
```cpp
auto hasConstantOnRight = [](Expression*) {
  // check if binary, and if there is a constant on the right;
};
if (hasConstantOnRight(binary->left) && !hasConstantOnRight(binary->right)) swap
```

Reply:
It's great idea. The only thing I decided to make a richer state for the result. So if `hasConstantOnRight` found special constants like `0` or `-1` it return state which tell us force swapping in any case.